executor/memtable_reader.go

```diff
@@ -696,3 +701,299 @@ func (e *clusterLogRetriever) close() error {
+					resp, err := util.InternalHTTPClient().Do(req)
+					if err != nil {
+						ch <- hotRegionsResult{err: errors.Trace(err)}
+						return
+					}
+					defer func() {
+						terror.Log(resp.Body.Close())
+					}()
+					if resp.StatusCode != http.StatusOK {
+						ch <- hotRegionsResult{err: errors.Errorf("request %s failed: %s", url, resp.Status)}
+						return
+					}
+					var historyHotRegions HistoryHotRegions
+					if err = json.NewDecoder(resp.Body).Decode(&historyHotRegions); err != nil {
+						ch <- hotRegionsResult{err: errors.Trace(err)}
+						return
+					}
+					ch <- hotRegionsResult{addr: address, messages: &historyHotRegions}
+				}, nil)
+			}(ch, srv.StatusAddr, body)
+		}
+	}
+	return results, nil
+}
+
+func (e *hotRegionsHistoryRetriver) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.extractor.SkipRequest || e.isDrained {
+		return nil, nil
+	}
+
+	if !e.retrieving {
+		e.retrieving = true
+		results, err := e.initialize(ctx, sctx)
+		if err != nil {
+			e.isDrained = true
+			return nil, err
+		}
+		// Initialize the heap
+		e.heap = &hotRegionsResponseHeap{}
+		for _, ch := range results {
+			result := <-ch
+			if result.err != nil {
+				sctx.GetSessionVars().StmtCtx.AppendWarning(result.err)
+			}
+			if result.err != nil || len(result.messages.HistoryHotRegion) == 0 {
+				continue
+			}
+			*e.heap = append(*e.heap, result)
+		}
+		heap.Init(e.heap)
+	}
+	// Merge the results
+	var finalRows [][]types.Datum
+	allSchemas := sctx.GetInfoSchema().(infoschema.InfoSchema).AllSchemas()
+	tz := sctx.GetSessionVars().Location()
+	tikvStore, ok := sctx.GetStore().(helper.Storage)
+	if !ok {
+		return nil, errors.New("Information about hot region can be gotten only when the storage is TiKV")
+	}
+	tikvHelper := &helper.Helper{
+		Store:       tikvStore,
+		RegionCache: tikvStore.GetRegionCache(),
+	}
+	for e.heap.Len() > 0 && len(finalRows) < hotRegionsHistoryBatchSize {
+		minTimeItem := heap.Pop(e.heap).(hotRegionsResult)
+		row, err := e.getHotRegionRowWithSchemaInfo(minTimeItem.messages.HistoryHotRegion[0], tikvHelper, allSchemas, tz)
+		if err != nil {
+			return nil, err
+		}
+		if row != nil {
+			finalRows = append(finalRows, row)
+		}
+		minTimeItem.messages.HistoryHotRegion = minTimeItem.messages.HistoryHotRegion[1:]
+		// Fetch next message item
+		if len(minTimeItem.messages.HistoryHotRegion) != 0 {
+			heap.Push(e.heap, minTimeItem)
+		}
+	}
+	// All streams are drained
+	e.isDrained = e.heap.Len() == 0
+	return finalRows, nil
+}
+
+func (e *hotRegionsHistoryRetriver) getHotRegionRowWithSchemaInfo(
+	hisHotRegion *HistoryHotRegion,
+	tikvHelper *helper.Helper,
+	allSchemas []*model.DBInfo,
+	tz *time.Location,
+) ([]types.Datum, error) {
+	_, startKey, _ := codec.DecodeBytes(hisHotRegion.StartKey, []byte{})
+	_, endKey, _ := codec.DecodeBytes(hisHotRegion.EndKey, []byte{})
+	region := &tikv.KeyLocation{StartKey: startKey, EndKey: endKey}
```

Comment:
Is there possible the old table is dropped?

Reply:
yes, we have discussed this case, dropped tables' hot region will be ignore in current implement.
```go
	_, startKey, _ := codec.DecodeBytes(hisHotRegion.StartKey, []byte{})
	_, endKey, _ := codec.DecodeBytes(hisHotRegion.EndKey, []byte{})
	region := &tikv.KeyLocation{StartKey: startKey, EndKey: endKey}
	hotRange, err := helper.NewRegionFrameRange(region)
	if err != nil {
		return nil, err
	}

	f := tikvHelper.FindTableIndexOfRegion(allSchemas, hotRange)
	// Ignore row without corresponding schema f.
	if f == nil {
		return nil, nil
	}
```
This case has been tested in `memtable_reader_test.go/TestTiDBHotRegionsHistory` line 1097-1099, code run as expected.
```go
		//      table_id = 131, index_id = 1, index_value = 1, deleted schema
		{"2019-10-10 10:10:23", "UNKONW", "UNKONW", "131", "UNKONW", "1", "5", "5", "55555", "0", "1", "READ", "99", "99", "99", "99"},
		{"2019-10-10 10:10:24", "UNKONW", "UNKONW", "131", "UNKONW", "1", "6", "6", "66666", "0", "0", "WRITE", "99", "99", "99", "99"},
```